onnxruntime/core/framework/bfc_arena.cc

```diff
@@ -95,30 +87,54 @@ bool BFCArena::Extend(size_t rounded_bytes) {
     return new_mem;
   };
 
-  void* mem_addr = safe_alloc(bytes);
-
-  if (mem_addr == nullptr) {
-    static constexpr float kBackpedalFactor = 0.9f;
+  auto get_extend_bytes = [this, available_bytes](const size_t bytes) -> size_t {
+    size_t extend_bytes = 0;
+    if (arena_extend_strategy_ == ArenaExtendStrategy::kNextPowerOfTwo) {
+      // If curr_region_allocation_bytes_ is not enough to satisfy the
+      // allocation, keep multiplying by a power of two until that is
+      // sufficient.
+      bool increased_allocation = false;
+      while (bytes > curr_region_allocation_bytes_) {
+        curr_region_allocation_bytes_ *= 2;
+        increased_allocation = true;
+      }
 
-    // Try allocating less memory.
-    while (mem_addr == nullptr) {
-      bytes = RoundedBytes(static_cast<size_t>(bytes * kBackpedalFactor));
-      if (bytes < rounded_bytes)
-        break;
+      extend_bytes = std::min(curr_region_allocation_bytes_, available_bytes);
 
-      mem_addr = safe_alloc(bytes);
+      // we allocated the same number of bytes as the current region, so we have 2x that now
+      if (!increased_allocation) {
```

Comment:
This comment is actually slightly misleading (was in the old code too). The 2x is to double the minimum size of the next amount we'll allocate rather than reflecting the current total.